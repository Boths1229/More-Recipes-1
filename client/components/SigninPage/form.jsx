/* eslint-disable no-unused-vars */
import React, { PropTypes } from 'react';
import img from '../../public/images/burger-image.jpg';

const propTypes = {
  error: PropTypes.shape({
    error: PropTypes.string,
  }),
  onSubmit: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
};

const defaultProps = {
  error: null
};

const Form = props => (
  <div className="container">
    <div className="pt-5" style={{ textAlign: 'center' }}>
      <form className="m-x-auto form card" onSubmit={props.onSubmit}>
        <h1 className="signin-text"> More-Recipes</h1>
        <div>
          <img
            src={img}
            name="image"
            className="img-fluid rounded-circle img-size"
            alt="logo"
          />
        </div>
        {
          props.error.error &&
          <div className="alert alert-danger">
            {props.error.error}
          </div>
        }
        <div className="pb-4 pt-3 form-group">
          <input
            type="text"
            name="username"
            onFocus={props.onFocus}
            onChange={props.onChange}
            className="form-control"
            required
            placeholder="Username or Email"
          />
        </div>
        <div className="pb-4 form-group">
          <input
            type="password"
            name="password"
            onFocus={props.onFocus}
            onChange={props.onChange}
            className="form-control"
            required
            placeholder="Password"
          />
        </div>
        <button
          type="submit"
          className="btn signin-btn btn-lg"
          style={{ color: '#f2b43c' }}
        >
          Log in
        </button>
        <p> Donâ€™t have an account?
          <a href="/signup">Create an account.</a>
        </p>
      </form>
    </div>
  </div>
);

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

export default Form;
