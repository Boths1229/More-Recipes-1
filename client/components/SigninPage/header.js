import React from 'react';

/* eslint-disable no-unused-vars */
const Header = (props) => {
  return (
    <header>
      <nav className="navbar navbar-expand-lg navbar-dark" style={{backgroundColor: "#2b3034"}}>
        <a className="navbar-brand" href="/">More-Recipes</a>
        <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
          <span className="navbar-toggler-icon"></span>
        </button>

        <div className="collapse navbar-collapse" id="navbarSupportedContent">
          <ul className="navbar-nav ml-auto">
            <li className="nav-item">
              <a className="nav-link" href="/signup">Signup</a>
            </li>
          </ul>
        </div>
      </nav>
    </header>
  );
}

export default Header;