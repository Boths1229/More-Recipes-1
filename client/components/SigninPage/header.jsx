/* eslint-disable no-unused-vars */
import React from 'react';
import { Link } from 'react-router';


const Header = (props) => {
  return (
    <header>
      <nav className="navbar navbar-expand-lg navbar-dark" style={{ backgroundColor: '#2b3034' }}>
        <a className="navbar-brand" href="/">More-Recipes</a>
        <button
          className="navbar-toggler"
          type="button"
          data-toggle="collapse"
          data-target="#navbarSupportedContent"
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation">
          <span className="navbar-toggler-icon"></span>
        </button>

        <div className="collapse navbar-collapse" id="navbarSupportedContent">
          <ul className="navbar-nav ml-auto">
            <li className="nav-item">
              <Link className="nav-link" to="/signup">Signup</Link>
            </li>
          </ul>
        </div>
      </nav>
    </header>
  );
}

export default Header;