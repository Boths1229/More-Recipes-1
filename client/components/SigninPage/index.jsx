import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import toastr from 'toastr';
import Header from './header.jsx';
import Form from './form.jsx';
import Footer from '../common/footer.jsx';
import img from '../../public/images/banner-img-2.jpg';
import signinAction from '../../actionsCreator/signin';
import { resetUserError } from '../../actionsCreator/signup';

/* eslint-disable react/no-unused-state */

const propTypes = {
  authenticated: PropTypes.bool.isRequired,
  error: PropTypes.string,
  actions: PropTypes.shape({
    resetUserError: PropTypes.func,
    signinAction: PropTypes.func,
  }).isRequired
};

const defaultProps = {
  error: null,
};

/**
 * @description A class to display the signin page
 *
 * @extends Component
 */
class SigninPage extends Component {
  /**
   * enables action to the performed on this component
   *
   * @param {object} props
   * @param {object} context
  */
  constructor(props, context) {
    super(props, context);
    this.state = {
      username: '',
      email: '',
      password: '',
      error: '',
    };
    this.onChange = this.onChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.onFocus = this.onFocus.bind(this);
  }

  /**
   * @description performs DOM manipulation right before the component mount
   *
   * @memberof SigninPage
   *
   * @returns {void}
  */
  componentWillMount() {
    document.body.style.backgroundImage = `url(${img}`;
    if (this.props.authenticated === true) {
      this.context.router.push('/recipes');
    }
  }

  /**
   * @description performs an action when component receives props
   *
   * @memberof SigninPage
   *
   * @param {object} nextProps
   *
   * @returns {void}
  */
  componentWillReceiveProps(nextProps) {
    const { authenticated, error } = nextProps;
    if (authenticated === true) {
      toastr.success('Login successfull');
      this.context.router.push('/recipes');
    } else {
      this.setState({
        error,
      });
    }
  }


  /**
   * @description performs action after the component unmount
   *
   * @memberof SigninPage
   *
   * @returns {void}
  */
  componentWillUnmount() {
    document.body.style.backgroundImage = '';
  }

  /**
   * @description listen for changes in the form
   *
   * @param {object} event
   *
   * @memberof SigninPage
   *
   * @returns {void}
  */
  onChange(event) {
    const { name } = event.target,
      { value } = event.target;
    this.setState({
      [name]: value
    });
  }

  /**
   * @description removes errorMessage when focused
   *
   * @param {object} event
   *
   * @memberof SigninPage
   *
   * @returns {void}
  */
  onFocus(event) {
    const { name } = event.target;
    const error = '';
    switch (name) {
      case 'username':
        this.setState({ error: '' });
        this.props.actions.resetUserError(error);
        break;
      case 'password':
        this.setState({ error: '' });
        this.props.actions.resetUserError(error);
        break;
      default:
        return this.state;
    }
  }

  /**
   * @description handles the submission of form
   *
   * @param {object} event
   *
   * @memberof SigninPage
   *
   * @returns {void}
  */
  handleSubmit(event) {
    event.preventDefault();
    this.props.actions.signinAction(this.state);
  }

  /**
   * @description renders the components
   *
   * @memberof SigninPage
   *
   * @returns {void} returns the components
  */
  render() {
    return (
      <div>
        <Header />
        <Form
          error={this.state}
          onFocus={this.onFocus}
          onChange={this.onChange}
          onSubmit={this.handleSubmit}
        />
        <Footer />
      </div>
    );
  }
}

SigninPage.contextTypes = {
  router: PropTypes.object
};

SigninPage.propTypes = propTypes;
SigninPage.defaultProps = defaultProps;

/**
 * @description maps state to properties of SigninPage
 *
 * @param  {object} state
 *
 * @returns {object} returns the state to be mapped to props
 */
function mapStateToProps(state) {
  return {
    authenticated: state.auth.authenticated,
    error: state.auth.error
  };
}

/**
 * @description maps action to properties of SigninPage
 *
 * @param  {object} dispatch
 *
 * @returns {object} returns the action to be bind
 */
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      signinAction,
      resetUserError
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(SigninPage);
