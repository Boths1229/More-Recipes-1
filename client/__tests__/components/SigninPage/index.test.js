import React from 'react';
import { mount } from 'enzyme';
import { SigninPage } from '../../../components/SigninPage/index';

describe('Signin component', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      authenticated: '',
      error: '',
      actions: {
        resetUserError: () => {},
        signinAction: () => {}
      }
    };
    wrapper = mount(<SigninPage {...props} />);
  });

  it('should  render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set username state when username input value changes', () => {
    const event = {
      target: { name: 'username', value: '' }
    };

    event.target.value = 'tywo';
    wrapper.find('input').at(0).simulate('change', event);
    expect(wrapper.state('username')).toBe('tywo');
  });

  it('should set error when username input is focused on', () => {
    wrapper.setState({ error: 'user not found' });

    const event = {
      target: { name: 'username', value: '' }
    };

    wrapper.find('input').at(0).simulate('focus', event);
    expect(wrapper.state('error')).toBe('');
  });

  it('should set error when password input is focused on', () => {
    wrapper.setState({ error: 'invalid details' });

    const event = {
      target: { name: 'password', value: '' }
    };

    wrapper.find('input').at(1).simulate('focus', event);
    expect(wrapper.state('error')).toBe('');
  });

  it('should return  state when no input is on focused', () => {
    wrapper.setState({ error: 'invalid details' });

    const event = {
      target: { name: '', value: '' }
    };

    wrapper.find('input').at(0).simulate('focus', event);
    expect(wrapper.state('error')).toBe('invalid details');
  });

  it('should signin user when form is submitterd', () => {
    const action = wrapper.instance();
    const signin = jest.spyOn(wrapper.instance(), 'handleSubmit');
    action.handleSubmit({ preventDefault: () => {} });
    expect(signin).toBeCalled();
  });
});
