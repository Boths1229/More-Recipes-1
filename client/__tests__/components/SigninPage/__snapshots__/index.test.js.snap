// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin component should  render correctly 1`] = `
<SigninPage
  actions={
    Object {
      "resetUserError": [Function],
      "signinAction": [Function],
    }
  }
  authenticated=""
  error=""
>
  <div>
    <Header>
      <header>
        <nav
          className="navbar navbar-expand-lg navbar-dark"
          style={
            Object {
              "backgroundColor": "#2b3034",
            }
          }
        >
          <a
            className="navbar-brand"
            href="/"
          >
            More-Recipes
          </a>
          <button
            aria-controls="navbarSupportedContent"
            aria-expanded="false"
            aria-label="Toggle navigation"
            className="navbar-toggler"
            data-target="#navbarSupportedContent"
            data-toggle="collapse"
            type="button"
          >
            <span
              className="navbar-toggler-icon"
            />
          </button>
          <div
            className="collapse navbar-collapse"
            id="navbarSupportedContent"
          >
            <ul
              className="navbar-nav ml-auto"
            >
              <li
                className="nav-item"
              >
                <Link
                  className="nav-link"
                  onlyActiveOnIndex={false}
                  style={Object {}}
                  to="/signup"
                >
                  <a
                    className="nav-link"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    Signup
                  </a>
                </Link>
              </li>
            </ul>
          </div>
        </nav>
      </header>
    </Header>
    <Form
      error={
        Object {
          "email": "",
          "error": "",
          "password": "",
          "username": "",
        }
      }
      onChange={[Function]}
      onFocus={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="container"
      >
        <div
          className="pt-5"
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          <form
            className="m-x-auto form card"
            onSubmit={[Function]}
          >
            <h1
              className="signin-text"
            >
               More-Recipes
            </h1>
            <div>
              <img
                alt="logo"
                className="img-fluid rounded-circle img-size"
                name="image"
                src="test"
              />
            </div>
            <div
              className="pb-4 pt-3 form-group"
            >
              <input
                className="form-control"
                name="username"
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Username or Email"
                required={true}
                type="text"
              />
            </div>
            <div
              className="pb-4 form-group"
            >
              <input
                className="form-control"
                name="password"
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Password"
                required={true}
                type="password"
              />
            </div>
            <button
              className="btn signin-btn btn-lg"
              style={
                Object {
                  "color": "#f2b43c",
                }
              }
              type="submit"
            >
              Log in
            </button>
            <p>
               Don’t have an account?
              <a
                href="/signup"
              >
                Create an account.
              </a>
            </p>
          </form>
        </div>
      </div>
    </Form>
    <Footer>
      <footer
        className="footer-section"
        id="footer-below"
      >
        <div
          className="container"
        >
          <strong
            className="copyright"
          >
            More-Recipes © 2018, All Rights Reserved, Design & Developed By: SannyTee
          </strong>
        </div>
      </footer>
    </Footer>
  </div>
</SigninPage>
`;
